File: server/routes/auth_routes.py
"""
Authentication routes
Handles login, logout, and session management
"""
from fastapi import APIRouter, Request, HTTPException, status
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

from config import settings
from services.angel_auth_service import angel_auth_service

router = APIRouter()


class LoginRequest(BaseModel):
    """Login request model"""
    clientId: str
    pin: str
    totpSecret: str
    apiKey: str
    apiSecret: Optional[str] = None


@router.get("/defaults")
async def get_defaults():
    """Get default login credentials from environment"""
    return {
        "success": True,
        "defaults": settings.default_login
    }


@router.post("/login")
async def login(request: Request, login_data: LoginRequest):
    """
    Login to Angel One with credentials.
    
    Args:
        request: FastAPI request object
        login_data: Login credentials
        
    Returns:
        Success response with client ID
    """
    # Validate required fields
    if not all([login_data.clientId, login_data.pin, login_data.totpSecret, login_data.apiKey]):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Missing required field(s): clientId, pin, totpSecret, apiKey"
        )
    
    try:
        # Attempt login
        tokens = await angel_auth_service.login_with_credentials(
            login_data.clientId,
            login_data.pin,
            login_data.totpSecret,
            login_data.apiKey
        )
        
        # Store in session (server-side only, never sent to browser)
        request.session["angel"] = {
            "clientId": login_data.clientId,
            "pin": login_data.pin,
            "totpSecret": login_data.totpSecret,
            "apiKey": login_data.apiKey,
            "apiSecret": login_data.apiSecret,
            **tokens,
            "loggedInAt": datetime.utcnow().isoformat() + "Z"
        }
        
        return {
            "success": True,
            "clientId": login_data.clientId
        }
    
    except Exception as err:
        error_message = str(err) or "Login failed"
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail=error_message
        )


@router.post("/logout")
async def logout(request: Request):
    """
    Logout from Angel One and clear session.
    
    Args:
        request: FastAPI request object
        
    Returns:
        Success response
    """
    angel = request.session.get("angel")
    
    try:
        if angel:
            # Best-effort logout from Angel One
            await angel_auth_service.logout(
                angel["apiKey"],
                angel["jwtToken"],
                angel["clientId"]
            )
    except Exception:
        pass  # Clear local session regardless
    finally:
        # Clear session
        request.session.clear()
        
        return {"success": True}


@router.get("/session")
async def get_session(request: Request):
    """
    Get current session status.
    
    Args:
        request: FastAPI request object
        
    Returns:
        Session status with authentication state
    """
    angel = request.session.get("angel")
    
    return {
        "authenticated": bool(angel and angel.get("jwtToken")),
        "clientId": angel.get("clientId") if angel else None
    }

# Made with Bob

File: server/src/routes/auth.routes.js
const express = require('express');
const config = require('../config');
const { loginWithCredentials, logout } = require('../services/angelAuth.service');

const router = express.Router();

router.get('/defaults', (req, res) => {
  res.json({
    success: true,
    defaults: config.defaultLogin,
  });
});

router.post('/login', async (req, res) => {
  const { clientId, pin, totpSecret, apiKey, apiSecret } = req.body || {};

  const missing = ['clientId', 'pin', 'totpSecret', 'apiKey'].filter((f) => !req.body || !req.body[f]);
  if (missing.length) {
    return res.status(400).json({ success: false, message: `Missing required field(s): ${missing.join(', ')}` });
  }

  try {
    const tokens = await loginWithCredentials({ clientId, pin, totpSecret, apiKey });

    // Kept server-side only, in memory, for this session - never sent back
    // to the browser. pin/totpSecret are retained so we can transparently
    // re-login if the refresh token also expires (Angel One sessions die at
    // midnight regardless of activity).
    req.session.angel = {
      clientId,
      pin,
      totpSecret,
      apiKey,
      apiSecret: apiSecret || null,
      ...tokens,
      loggedInAt: new Date().toISOString(),
    };

    req.session.save(() => {
      res.json({ success: true, clientId });
    });
  } catch (err) {
    const message = err.response?.data?.message || err.message || 'Login failed';
    res.status(401).json({ success: false, message });
  }
});

router.post('/logout', async (req, res) => {
  const angel = req.session && req.session.angel;
  try {
    if (angel) {
      await logout(angel).catch(() => {}); // best-effort; clear local session regardless
    }
  } finally {
    req.session.destroy(() => {
      res.clearCookie('connect.sid');
      res.json({ success: true });
    });
  }
});

router.get('/session', (req, res) => {
  const angel = req.session && req.session.angel;
  res.json({
    authenticated: !!(angel && angel.jwtToken),
    clientId: angel ? angel.clientId : null,
  });
});

module.exports = router;
